### DISCUSSION TITLE
API Key rotation 시스템 개발

### DISCUSSION THREAD

**Discussion by JuneParkCode** [manito42/backend#101] [General]
## 문제
- 42 API Secret key의 경우 일정 주기로 무호화되며 새로운 key로 변경해야합니다.
- 따라서 새로운 key를 받아오고 제 때 교체를 할 수 있는 시스템이 필요합니다.

## 분석
- secret key의 경우 일정 주기로 무효화
- 다음 secret key를 생성할 수 있으며 해당 key는 이전 secret key가 무효화 되는 시점에 교체됨
  - 이는 미리 generate 버튼을 통해 생성해둬야함.
- API Key를 받기 위해서는 intra 에 직접 접속하여 획득해야할 수 있음.
  - 별도로 API 다음 키를 받는 방안은 존재하지 않음.
  - 자동화를 하기 위해서는 크롤러를 이용해야함
  - 자동화 앱에서 결국 운영자의 42 ID/PW 주입이 필요하기 때문에 내키지는 않음.

## 해결 방안
- 현재 key와 next key를 APP에 전달함
- invalidate 되는 시점을 명확하게 명시하여 그 시점 이전에 next key로 변경될 수 있도록 해야함
  - 개발자 편의성을 위해서 관련 notification을 주는 것도 좋을 것 같음
- config 에서 값을 직접 가져오는 현재의 시스템에서 변경될 필요가 있음
  - env 에서 주입 받고, 시간에 따라 key를 관리하는 별도의 장치가 필요함
- 키 관리를 하는 장치를 개발할 때, 여러 키를 rotation 하면서 사용할 수 있도록 하는 것도 좋을 것 같음
  - api key limit 제한 때문
  - 여러개의 Key를 받아와 사용.
  - 주입 방법이 env 가 아닌 다른 방식을 생각해봐야함
  - 단순히 csv 파일에서 파싱해오거나, 별도의 저장소를 두거나...
- auth 관련한 별도의 서버를 구성하는 것도 하나의 방법임
  - Java + Spring 으로 이동할 때 이걸 별도로 구성해서 제공해보는 것도 생각해볼만함

@koreanddinghwan

**Comment by koreanddinghwan** [comment]
**테스트용으로 생성한 API key입니다.**

<img width="624" alt="스크린샷 2023-09-20 11 41 21" src="https://github.com/manito42/backend/assets/76278794/42901a95-3115-4896-8446-213ce68300dc">

- 생성된 key의 유효기간은 **Replace/Generate**된 날짜 기준 **28일**뒤
- 생성한 날짜 기준 **21일 뒤**, **NEXT SECRET**이 자동생성됨

**Comment by JuneParkCode** [comment]
제일 관건은 key를 어떤 방식으로 저장하는가일 것 같은데요
별도의 DB table을 둬서 관리할까요? 그런데 이것대로 매 로그인마다 DB 에서 API 키를 받아오는 것도 적합하지는 않은 것 같다는 생각이 드네요.

**Comment by koreanddinghwan** [comment]
<img width="483" alt="스크린샷 2023-09-20 13 15 36" src="https://github.com/manito42/backend/assets/76278794/c0dd270b-effa-44d2-b836-130ae4cf7cee">

Event Bridge로 스케쥴링이 가능해서...  
크롤링 방식으로 처리하면 완전히 자동화가 가능하긴합니다.

**Reply by koreanddinghwan** [reply]
Secret을 어떻게 들고있어야할지는 저도 고민되긴하네요..

**Reply by koreanddinghwan** [reply]
Database에 들고있지 않게하려면 결국, In-memory로 들고있어야하는데,

현재 켜져있는 서버에 **별도의 Admin용 API를 생성해** configService.set을 호출해 memory의 secret값을 바꿀 수 있습니다

**Reply by JuneParkCode** [reply]
in memory 방법의 경우 instance 가 여러 개 존재하는 경우가 문제가 될 수 있을 것 같아요.

**Reply by koreanddinghwan** [reply]
맞네요 , 그럼 in memory로 한다고하면 polling으로 서버에서 계속 secret이 최신인지 확인해야하는거죠/ 아니면 하루에 1번으로 빈도수를 낮춰놓던가

**Reply by koreanddinghwan** [reply]
저거대로하면, 21일차에 NEXT_TOKEN이 생성되고, 25일차에 람다가 작동해서 Secret Manager안에있는 Secret값을 바꾸는데, 이와 동시에 기존 Secret은 무효화되면서 공백기간동안 회원가입이 안되겠네요

**Reply by koreanddinghwan** [reply]
Secret을 바꾼다고하면 결국 데이터베이스를 사용해야할 것 같습니다

**Reply by JuneParkCode** [reply]
자동화 한다면 
Secret key의 유효 기간을 적어도 1일 이상 차이나도록 여러개를 둡니다. 그리고 DB에 Available list 에는 유효한 키만 유지

변경 과정에서는
기존 키 DB에서 우선 제거
키 replace
키 available list 에 업데이트

이렇게 하면 available list 에 사용할 수 있는 키가 있으니 그거 사용해서 문제없이 갈 수 있습니다.
대신 이럴 경우 로그인마다 secret db를 뒤져서 가져와야하는게 문제가 될 수 있습니다.

그리고 rotation 하면서 사용해야하는데, db에서 ramdom 하게 들고오는 것도 좀 골치아프고...

**Reply by koreanddinghwan** [reply]
최소한 2개는 필요하겠네요.

**Reply by koreanddinghwan** [reply]
random하게 가져오는것도 나쁘지 않아보이는데요? 

|pk|uid|secret| 이런식으로해서 pk로 가져와도..?

**Reply by JuneParkCode** [reply]
로그인마다 db 접근 부분은 어떻게 생각하시는지?

**Reply by koreanddinghwan** [reply]
일반적인 경우에는 이렇게되면 In-memory로 레디스를 쓰던가해야하는데, 비싸..죠?

레디스를 쓰는 것에 대한 차선책으로 사용중인 관계형 데이터베이스를 사용해야합니다.(단, 반드시 클러스터링 인덱스 사용해서 딜레이 최소화)

- 토큰을 일부러 삭제하지 않는 이상 토큰 유효기간 동안 DB접근은 없고
- 토큰을 일부러 삭제한다면, 삭제함으로서 발생하는 로그인 딜레이(DB접근-> Secret가져와서 로그인처리)는 귀책사유가 클라이언트에게 있다고 생각합니다.

**Reply by JuneParkCode** [reply]
오케이 그러면 DB를 통해서 Secret 관리하고, 유효기간 넘어가는 토큰 접근 시 새로 받아오는걸로?
